org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/util/SoftClassCache.java

```diff
@@ -0,0 +1,85 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Christoph Läubrich.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.internal.compiler.util;
+
+import java.io.IOException;
+import java.lang.ref.SoftReference;
+import java.nio.file.Path;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Implements a soft cache for reading class files from disk, as these caches can grow quite large but data can be
+ * recovered afterwards we only hold a soft reference to the bytes itself.
+ */
+class SoftClassCache {
+
+	private ConcurrentMap<String, JdkClasses> jdks = new ConcurrentHashMap<>(2 ^ 4);
+
+	void clear() {
+		this.jdks.clear();
+	}
+
+	public byte[] getClassBytes(Jdk jdk, Path path) throws IOException {
+		return this.jdks.computeIfAbsent(jdk.path, JdkClasses::new).get(path);
+	}
+
+	private static final class JdkClasses {
+		private ConcurrentMap<Path, ClassBytes> classes = new ConcurrentHashMap<>(2 ^ 15);
```

Comment:
`2 ^ 15` == 13. Did you meant `2 << 15` (65536)? Why not just keep as it was (10007)?

I would prefer not to address different issues in a single commit. If you've observed issues with performance due different cache size, please report an issue.

Reply:
> 2 ^ 15 == 13. Did you meant 2 << 15 (65536)? Why not just keep as it was (10007)?

2 ^ 15 = 32768 and I choose that because in the debugger I see that at most there are about 25.000 classes cached, and I have not found any explanation why `10007` should be a good default, even if the JVM might be smart enough it seem to not very well align on usual byte boundaries that are multiple of 2 ...

So currently the cache grows from 10007 > 20014 > 40028 with the internal data structure, now it currently never resizes, but again here compared to the size of the bytes from the classes it should not matter much.

Reply:
@laeubi  for future reference:
```
jshell> 2^4
$1 ==> 6

jshell> 2 ^15
$2 ==> 13

jshell> 2 << 4
$3 ==> 32

jshell> 2 << 15
$4 ==> 65536
```